File: vinyl-catalog/src-tauri/src/commands/spotify.rs
use serde::{Deserialize, Serialize};

use crate::commands::keyring;

#[derive(Serialize)]
pub struct SpotifySearchResult {
    #[serde(rename = "albumId")]
    pub album_id: String,
    #[serde(rename = "albumUrl")]
    pub album_url: Option<String>,
    #[serde(rename = "previewUrl")]
    pub preview_url: Option<String>,
}

#[derive(Deserialize)]
struct SpotifyTokenResponse {
    access_token: String,
}

#[derive(Deserialize)]
struct SpotifyImage {
    url: String,
}

#[derive(Deserialize)]
struct SpotifyAlbumExternalUrls {
    spotify: Option<String>,
}

#[derive(Deserialize)]
struct SpotifyAlbum {
    id: String,
    external_urls: Option<SpotifyAlbumExternalUrls>,
    images: Option<Vec<SpotifyImage>>,
}

#[derive(Deserialize)]
struct SpotifyAlbums {
    items: Vec<SpotifyAlbum>,
}

#[derive(Deserialize)]
struct SpotifySearchResponse {
    albums: Option<SpotifyAlbums>,
}

#[tauri::command]
pub async fn spotify_search(q: String) -> Result<SpotifySearchResult, String> {
    let client_id = keyring::get_api_key("spotify-client-id")
        .ok_or("Spotify client ID not configured. Use save_api_key('spotify-client-id', '...') to set it.")?;
    let client_secret = keyring::get_api_key("spotify-client-secret")
        .ok_or("Spotify client secret not configured. Use save_api_key('spotify-client-secret', '...') to set it.")?;

    let client = reqwest::Client::new();

    // Step 1: Get access token via client credentials
    use base64::{engine::general_purpose::STANDARD, Engine};
    let credentials = STANDARD.encode(format!("{}:{}", client_id, client_secret));

    let token_resp: SpotifyTokenResponse = client
        .post("https://accounts.spotify.com/api/token")
        .header("Authorization", format!("Basic {}", credentials))
        .header("Content-Type", "application/x-www-form-urlencoded")
        .body("grant_type=client_credentials")
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json()
        .await
        .map_err(|_| "Failed to authenticate with Spotify")?;

    // Step 2: Search for the album
    let search_url = format!(
        "https://api.spotify.com/v1/search?q={}&type=album&limit=1",
        urlencoding::encode(&q)
    );

    let search_resp: SpotifySearchResponse = client
        .get(&search_url)
        .header("Authorization", format!("Bearer {}", token_resp.access_token))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json()
        .await
        .map_err(|e| e.to_string())?;

    let album = search_resp
        .albums
        .and_then(|a| a.items.into_iter().next())
        .ok_or("No matching Spotify album found")?;

    Ok(SpotifySearchResult {
        album_id: album.id,
        album_url: album.external_urls.and_then(|u| u.spotify),
        preview_url: album.images.and_then(|imgs| imgs.into_iter().next().map(|i| i.url)),
    })
}

File: api/src/routes/spotify.ts
import { Router } from 'express'
import { logger } from '../logger'

const log = logger.child({ module: 'spotify' })

const router = Router()

let cachedToken: string | null = null
let tokenExpiresAt = 0

async function getSpotifyToken(): Promise<string> {
  const now = Date.now()
  if (cachedToken && now < tokenExpiresAt - 60_000) {
    return cachedToken
  }

  const clientId = process.env.SPOTIFY_CLIENT_ID
  const clientSecret = process.env.SPOTIFY_CLIENT_SECRET

  if (!clientId || !clientSecret) {
    throw new Error('SPOTIFY_CLIENT_ID and SPOTIFY_CLIENT_SECRET must be set in server environment')
  }

  const credentials = Buffer.from(`${clientId}:${clientSecret}`).toString('base64')
  const resp = await fetch('https://accounts.spotify.com/api/token', {
    method: 'POST',
    headers: {
      Authorization: `Basic ${credentials}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: 'grant_type=client_credentials',
  })

  if (!resp.ok) {
    throw new Error(`Spotify token request failed: HTTP ${resp.status}`)
  }

  const data = (await resp.json()) as { access_token: string; expires_in: number }
  cachedToken = data.access_token
  tokenExpiresAt = now + data.expires_in * 1000
  return cachedToken
}

/**
 * @swagger
 * /api/spotify/search:
 *   post:
 *     summary: Search Spotify for an album
 *     tags: [Spotify]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required:
 *               - q
 *             properties:
 *               q:
 *                 type: string
 *     responses:
 *       200:
 *         description: Spotify search result
 */
router.post('/search', async (req, res) => {
  const q = (req.body as Record<string, unknown>).q as string | undefined
  if (!q || typeof q !== 'string') {
    return res.status(400).json({ error: 'q (search query string) is required' })
  }
  if (q.length > 500) {
    return res.status(400).json({ error: 'q parameter too long' })
  }

  try {
    const token = await getSpotifyToken()

    const searchUrl = `https://api.spotify.com/v1/search?q=${encodeURIComponent(q)}&type=album&limit=1`
    const resp = await fetch(searchUrl, {
      headers: { Authorization: `Bearer ${token}` },
    })

    if (!resp.ok) {
      log.error({ status: resp.status }, 'Spotify search failed')
      return res.status(502).json({ error: 'Upstream Spotify search failed' })
    }

    const data = (await resp.json()) as {
      albums?: { items?: { id: string; external_urls?: { spotify?: string }; images?: { url: string }[] }[] }
    }
    const album = data.albums?.items?.[0]

    if (!album) {
      return res.json({ albumId: null, albumUrl: null, previewUrl: null })
    }

    res.json({
      albumId: album.id,
      albumUrl: album.external_urls?.spotify ?? null,
      previewUrl: album.images?.[0]?.url ?? null,
    })
  } catch (err) {
    log.error({ err }, 'Spotify search error')
    res.status(502).json({ error: 'Upstream service error' })
  }
})

export default router
